src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -344,7 +346,10 @@ public synchronized void add(R record, SinkRecord original, long timeoutMs) {
     }
 
     unsentRecords.addLast(record);
-    recordMap.put(record, original);
+    if (reporter != null) {
+      // avoid unnecessary operations if not using the reporter
+      recordsToReportOnError.put(record, original);
```

Comment:
Gating is required in every call. I suggest fleshing it out in a helper method. 
Example: 

```
  private void addRecordsToReport(R record, SinkRecord original) {
    if (reporter != null) {
      recordsToReportOnError.put(record, original);
    }
  }
```